docs/proposal/20211210-launch-templates-managedmachinepools.md

```diff
@@ -0,0 +1,259 @@
+
+### Goals
+
+- Consistent API to use launch templates for `AWSMachinePool` and `AWSManagedMachinePool`
+- Single point of reconciliation of launch templates
+- Guide to the deprecation of certain API elements in `AWSManagedMachinePool`
+
+### Non-Goals/Future Work
+
+- Add non-existent controller unit tests for `AWSMachinePool` and `AWSManagedMachinePool`
+
+## Proposal
+
+At a high level, the plan is to:
+
+1. Add a new `AWSLaunchTemplate` field to [AWSManagedMachinePoolSpec](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/exp/api/v1beta1/awsmanagedmachinepool_types.go#L65) that uses the existing [AWSLaunchTemplate](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/ec057ad6e613a6578f67bf68a6c77fbe772af933/exp/api/v1beta1/types.go#L58) struct. For example:
+
+```go
+// AWSLaunchTemplate specifies the launch template and version to use when an instance is launched. This field 
+// will become mandatory in the future and its recommended you use this over fields AMIType,AMIVersion,InstanceType,DiskSize,InstanceProfile.
+// +optional
+AWSLaunchTemplate AWSLaunchTemplate `json:"awsLaunchTemplate"`
+```
+
+2. Update the comments on the below fields of [AWSManagedMachinePoolSpec](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9bc29570614aa7123d79f042b6e6efc2aaf3e490/exp/api/v1beta1/awsmanagedmachinepool_types.go#L65) to indicate that the fields is deprecated and that `AWSlaunchTemplate` should be used.
+    - AMIVersion
+    - AMIType
+    - DiskSize
+    - InstanceType
+3. Add new `LaunchTemplateID` and `LaunchTemplateVersion` fields to [AWSManagedMachinePoolStatus](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9bc29570614aa7123d79f042b6e6efc2aaf3e490/exp/api/v1beta1/awsmanagedmachinepool_types.go#L171) to store details of the launch template and version used.
+4. Add a new `LaunchTemplateVersion` field to [AWSMachinePoolStatus](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/exp/api/v1beta1/awsmachinepool_types.go#L112) to store the version of the launch template used.
+5. [Refactor the code](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/ec057ad6e613a6578f67bf68a6c77fbe772af933/exp/controllers/awsmachinepool_controller.go#L383) from the `AWSMachinePool` controller that reconciles `AWSLaunchTemplate` into a common location so that it can be shared.
+6. Update the controller for `AWSManagedMachinePool` to use the `AWSLaunchTemplate` reconciliation logic.
+7. Add checks in the `AWSManagedMachinePool` create/update validation webhooks that stops users specifying `AWSLaunchTemplate` if fields `AMIType,AMIVersion,InstanceType,DiskSize,InstanceProfile` are set
+8. Add warning logs to the `AWSManagedMachinePool` create/update validation webhooks if fields `AMIType,AMIVersion,InstanceType,DiskSize,InstanceProfile` stating that these fields will be deprecated in the future and that `AWSLaunchTemplate` should be used instead
```

Comment:
Adding discussion from https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094#issuecomment-1227585754.

Let's decide whether we want to deprecate these fields or not.

Reply:
I've been thinking about this and think that its confusing to have 2 ways to specify the configuration for a machine pool.

@pydctw @richardchen331 @sedefsavas - i would say that we deprecate these fields still. wdyt?

Reply:
After thinking some more, +1 to deprecate these fields. 

Clarifying questions
* My understanding from the [comment](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094/files#r944742714) was that `AMIType` is still needed when using LaunchTemplate. Is this correct?
* `IamInstanceProfile` is a field in LaunchTemplate and I don't think it was in AWSManagedMachinePool spec, meaning it is not a field to deprecate. If I read the doc correctly, IAM InstanceProfile is prohibited in LaunchTemplate for EKS, the check is already implemented in the [webhook](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3094/files#diff-280ce5df0a5f3c77ecd44a29fb7fa50822c9be39b69385b58e5c820286a2d38cR133-R135)

Reply:
I'm ok with deprecating the fields. 

We still need to pass a AMIType of `CUSTOM` to when making calls to EKS using the [AWS SDK](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/eks#CreateNodegroupInput). We could hardcode the value though. Please let me know what you think. Thanks!